lib/Object/ObjectLinker.cpp

```diff
@@ -3470,6 +3470,18 @@ bool ObjectLinker::readAndProcessInput(Input *Input, bool IsPostLto) {
     if (layoutInfo)
       layoutInfo->recordInputKind(CurInput->getKind());
     bool ELFOverriddenWithBC = false;
+    if (!isBackendInitialized()) {
+      // Infer machine for selecting backend
+      eld::Expected<uint16_t> Machine =
+          getELFExecObjParser()->getMachine(*CurInput);
+      if (!Machine) {
+        ThisConfig.raiseDiagEntry(std::move(Machine.error()));
+        return false;
+      }
+      llvm::dyn_cast<ObjectFile>(CurInput)->setMachine(*Machine);
```

Comment:
Can we use llvm::cast if we are sure that cast would succeed? llvm::dyn_cast can return nullptr and using a nullptr might be an undefined behavior.